File: clients/mediafx_py/python/examples/shift.py
#!/usr/bin/env python
# Copyright (C) 2025 Andrew Wason
# SPDX-License-Identifier: GPL-3.0-or-later

from mediafx import MediaFX


def main() -> None:
    client = MediaFX()
    width, height = client.frame_size
    rendered_frame = bytearray(client.frame_bytecount)
    frames = [bytearray(client.frame_bytecount)]

    while True:
        (_, xshift, yshift, _) = client.render_frame(frames)

        xshift = int(xshift * width)
        yshift = int(yshift * height)
        for dy in range(0, height):
            for dx in range(0, width):
                sy = (dy + yshift) % height
                sx = (dx + xshift) % width
                dest_index = (dy * width + dx) * 4
                source_index = (sy * width + sx) * 4
                for channel in range(0, 4):
                    rendered_frame[dest_index + channel] = frames[0][source_index + channel]

        client.render_commit(rendered_frame)


if __name__ == "__main__":
    main()

File: clients/mediafx_node/examples/shift.ts
#!/usr/bin/env node
// Copyright (C) 2025 Andrew Wason
// SPDX-License-Identifier: GPL-3.0-or-later

const { MediaFX } = require("../index.js");

const client = new MediaFX();
const [width, height] = client.frameSize;
const renderedFrame = new Uint8Array(client.frameBytecount);
const frames = [new Uint8Array(client.frameBytecount)];

while (true) {
  let [, xshift, yshift] = client.renderFrame(frames);

  xshift = Math.floor(xshift * width);
  yshift = Math.floor(yshift * height);
  for (let dy = 0; dy < height; dy++) {
    for (let dx = 0; dx < width; dx++) {
      const sy = (dy + yshift) % height;
      const sx = (dx + xshift) % width;
      const destIndex = (dy * width + dx) * 4;
      const sourceIndex = (sy * width + sx) * 4;
      for (let channel = 0; channel < 4; channel++) {
        renderedFrame[destIndex + channel] = frames[0][sourceIndex + channel];
      }
    }
  }

  client.renderCommit(renderedFrame);
}
